great_expectations/core/config_provider.py

```diff
@@ -1,22 +1,364 @@
 from __future__ import annotations
 
+import base64
 import errno
+import json
+import logging
 import os
+import re
 from abc import ABC, abstractmethod
 from collections import OrderedDict
-from typing import Dict, Optional, Type, cast
+from functools import lru_cache
+from typing import Any, Dict, Optional, Type, cast
 
+import great_expectations.exceptions as ge_exceptions
 from great_expectations.core.yaml_handler import YAMLHandler
-from great_expectations.data_context.types.base import GeCloudConfig
-from great_expectations.data_context.util import (
-    substitute_all_config_variables,
-    substitute_config_variable,
-)
-
+from great_expectations.data_context.types.base import BaseYamlConfig, GeCloudConfig
+
+try:
+    from azure.identity import DefaultAzureCredential
+    from azure.keyvault.secrets import SecretClient
+except ImportError:
+    SecretClient = None
+    DefaultAzureCredential = None
+
+try:
+    import boto3
+    from botocore.exceptions import ClientError
+except ImportError:
+    boto3 = None
+    ClientError = None
+
+try:
+    from google.cloud import secretmanager
+except ImportError:
+    secretmanager = None
+
+logger = logging.getLogger(__name__)
 yaml = YAMLHandler()
 
 
+class ConfigurationSubstitutor:
+    def __init__(self) -> None:
+        # Using the @lru_cache decorator on method calls can create memory leaks - an attr is preferred here.
+        # Ref: https://stackoverflow.com/a/68550238
+        self._secret_store_cache = lru_cache(maxsize=None)(
+            self._substitute_value_from_secret_store
+        )
+
+    def substitute_all_config_variables(
+        self, data, replace_variables_dict, dollar_sign_escape_string: str = r"\$"
+    ):
```

Comment:
I can come back and type hint all of this (or let me know if you'd prefer I do that now)